File: app/src/main/java/com/biotel/iot/ui/main/recyclerview/binder/IBeaconBinder.java
package com.biotel.iot.ui.main.recyclerview.binder;

import android.content.Context;

import androidx.annotation.NonNull;

import com.biotel.iot.R;
import com.biotel.iot.ui.common.Navigation;
import com.biotel.iot.ui.common.recyclerview.BaseViewBinder;
import com.biotel.iot.ui.common.recyclerview.BaseViewHolder;
import com.biotel.iot.ui.common.recyclerview.RecyclerViewItem;
import com.biotel.iot.ui.main.recyclerview.holder.IBeaconHolder;

import uk.co.alt236.bluetoothlelib.device.beacon.ibeacon.IBeaconDevice;

import com.biotel.iot.ui.main.recyclerview.model.IBeaconItem;
import com.biotel.iot.util.Constants;

/**
 * Created By Piyush Pandey on 18-07-2020
 */
public class IBeaconBinder extends BaseViewBinder<IBeaconItem> {

    private final Navigation navigation;

    public IBeaconBinder(Context context, Navigation navigation) {
        super(context);
        this.navigation = navigation;
    }

    @Override
    public void bind(@NonNull BaseViewHolder<IBeaconItem> holder, @NonNull IBeaconItem item) {

        final IBeaconHolder actualHolder = (IBeaconHolder) holder;
        final IBeaconDevice device = item.getDevice();

        final String accuracy = Constants.DOUBLE_TWO_DIGIT_ACCURACY.format(device.getAccuracy());

        actualHolder.getIbeaconMajor().setText(String.valueOf(device.getMajor()));
        actualHolder.getIbeaconMinor().setText(String.valueOf(device.getMinor()));
        actualHolder.getIbeaconTxPower().setText(String.valueOf(device.getCalibratedTxPower()));
        actualHolder.getIbeaconUUID().setText(device.getUUID());
        actualHolder.getIbeaconDistance().setText(
                getContext().getString(R.string.formatter_meters, accuracy));
        actualHolder.getIbeaconDistanceDescriptor().setText(device.getDistanceDescriptor().toString());

        CommonBinding.bind(getContext(), actualHolder, device);
        actualHolder.getView().setOnClickListener(view -> navigation.openDetailsActivity(device));
    }

    @Override
    public boolean canBind(RecyclerViewItem item) {
        return item instanceof IBeaconItem;
    }
}

File: app/src/main/java/com/biotel/iot/ui/details/recyclerview/binder/IBeaconBinder.kt
package com.biotel.iot.ui.details.recyclerview.binder

import android.content.Context
import com.biotel.iot.R
import com.biotel.iot.ui.common.recyclerview.BaseViewBinder
import com.biotel.iot.ui.common.recyclerview.BaseViewHolder
import com.biotel.iot.ui.common.recyclerview.RecyclerViewItem
import com.biotel.iot.ui.details.recyclerview.holder.IBeaconHolder
import com.biotel.iot.ui.details.recyclerview.model.IBeaconItem
import com.biotel.iot.util.TimeFormatter
import uk.co.alt236.bluetoothlelib.resolvers.CompanyIdentifierResolver
import java.util.*

class IBeaconBinder(context: Context) : BaseViewBinder<IBeaconItem>(context) {

    override fun bind(holder: BaseViewHolder<IBeaconItem>, item: IBeaconItem) {
        val actualHolder = holder as IBeaconHolder
        val companyName = CompanyIdentifierResolver.getCompanyName(
                item.companyIdentifier,
                context.getString(R.string.unknown))
        actualHolder.companyId.text = getWithHexEncode(companyName, item.companyIdentifier)
        actualHolder.advert.text = getWithHexEncode(item.iBeaconAdvertisement)
        actualHolder.uuid.text = item.uuid
        actualHolder.major.text = getWithHexEncode(item.major)
        actualHolder.minor.text = getWithHexEncode(item.minor)
        actualHolder.txPower.text = getWithHexEncode(item.calibratedTxPower)
    }

    override fun canBind(item: RecyclerViewItem): Boolean {
        return item is IBeaconItem
    }

    companion object {
        private const val STRING_FORMAT = "%s (%s)"

        private fun getWithHexEncode(first: String, value: Int): String {
            return createLine(first, hexEncode(value))
        }

        private fun getWithHexEncode(value: Int): String {
            return createLine(value.toString(), hexEncode(value))
        }

        private fun createLine(first: String, second: String): String {
            return String.format(Locale.US, STRING_FORMAT, first, second)
        }

        private fun hexEncode(integer: Int): String {
            return "0x" + Integer.toHexString(integer).toUpperCase(Locale.US)
        }
    }
}